linode/terraform-provider-linode
# new: Add Placement Group fields to `linode_instance` resource

## 📝 Description

This pull request adds support for assigning a Linode to a Placement Group through a new `placement_group` field in the `linode_instance` resource, including the ability to reassign a Linode to a different PG. 

~~This PR relies on #1401 and will need to be rebased before being merged. In the meantime, relevant commits can be viewed here: https://github.com/linode/terraform-provider-linode/pull/1406/files/2a9f67928ea55f1c110ebc422259d5528a0358b0..HEAD~~

## ✔️ How to Test

The following test steps assume you have pulled down this PR locally and configured your environment to point to an API instance where Placement Groups are accessible:

```
export LINODE_URL=...
export LINODE_TOKEN=...
```

### Integration Testing

```
make PKG_NAME=linode/instance ARGS="-run TestAccResourceInstance_withPG" int-test
make PKG_NAME=linode/instance ARGS="-run TestAccResourceInstance_pgAssignment" int-test
```

### Manual Testing

1. In a Terraform provider sandbox environment (e.g. dx-devenv), apply the following configuration:

```terraform
resource "linode_instance" "my-instance" {
  label           = "my-instance"
  region          = linode_placement_group.test.region
  type            = "g6-standard-1"

  placement_group {
    id = linode_placement_group.test.id
  }
}

resource "linode_placement_group" "test" {
  label = "my-placement-group"
  region = "us-east"
  affinity_type = "anti_affinity:local"
}
```

2. Ensure the apply finishes successfully.
3. Navigate to the new instance & PG in Cloud Manager and ensure everything is configured as described in the configuration.
4. Re-plan the configuration to ensure the `label`, `is_strict`, and `affinity_type` fields have been implicitly computed/populated.
5. Apply the following configuration to unassign the Linode from the PG:

```terraform
resource "linode_instance" "my-instance" {
  label           = "my-instance"
  region          = linode_placement_group.test.region
  type            = "g6-standard-1"
}

resource "linode_placement_group" "test" {
  label = "my-placement-group"
  region = "us-east"
  affinity_type = "anti_affinity:local"
}
```

6. Ensure the apply finishes successfully.
7. Navigate to Cloud Manager and ensure the Linode has been unassigned from the PG.
